Remaining tweaks to middleware refactor + initial documentation changes for ``v7``

fselmo opened the pull request:
### What was wrong?

- Fix some inconsistencies in the middleware onion logic dealing with unhashable `Web3Builder.build` references.
- Remove references to middleware classes and opt for using the classes themselves.
- Allow turning off of ``HTTPProvider`` retries on exceptions if `exception_retry_configuration` is set to `None`; add test. Fix the default exception distinctions for sync vs async providers.
- Update the documentation for the middleware refactor, request retry configuration and caching.
- Started a ``v6`` -> ``v7`` migration guide.
- The ``AsyncHTTPProvider`` wasn't loading the right default exceptions for exception retry while I was documenting it. We also couldn't effectively turn off the retries for the sync provides. I went ahead and updated that in this PR as well and added the relevant test to make sure that is covered.
- ``normalize_request_parameters`` middleware was sitting in a stale state with no internal use. I removed it and updated the migration guide to reflect that.

Related to Issue #3169 

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![1000013739](https://github.com/ethereum/web3.py/assets/3532824/6dd4047e-3ddd-4eee-81d1-3a88d81f6ede)

fselmo committed:
Update documentation related to the middleware refactor:

- Add documentation for the new middleware model.
- Remove async versions of middleware from the docs; this is no longer
  a necessary distinction.
- Remove the `normalize_request_parameters` middleware from the docs and
  codebase. This wasn't used anywhere internally.

fselmo committed:
Update setup for HTTP rexception retry config

fselmo committed:
Update documentation on caching and exception retries for v7

fselmo committed:
Start the ``v6`` -> ``v7`` migration guide

fselmo committed:
Relevant newsfragments for #3211

fselmo committed:
Clarify how custom middleware can be added to the stack

- Clarify that middleware can be added to the onion either via the class
  or, in the case of the default middleware, a reference to the class.
- Minor typo fixes for docs updates.

fselmo committed:
Remove references to middleware classes; opt for class-only

fselmo committed:
Fix issues with missing repr of unhashables in middleware onion

fselmo committed:
Final cleanups after removing middleware references to classes

pacrob commented on docs/examples.rst:
```diff
@@ -688,8 +688,8 @@ Inject the middleware into the middleware onion
 
-    from web3.middleware import extradata_to_poa_middleware
-    w3.middleware_onion.inject(extradata_to_poa_middleware, layer=0)
+    from web3.middleware import ExtradataToPOAMiddleware
```
consistency nit: the error types all start `ExtraData...` and the field in question is `extraData`. What'chu think about changing the middleware name to `ExtraData...`?

pacrob commented on docs/v7_migration.rst:
```diff
@@ -0,0 +1,102 @@
+
+The middleware model has been changed to a class-based model. This allows for
+more flexibility in implementation, for asynchronous processing for example. Previously,
+the middleware were functions that tightly wrapped the `make_request` function of the
```
I think these function names should be double-ticks for code font, no?

pacrob commented on docs/internals.rst:
```diff
@@ -117,66 +117,187 @@ You can set a new list of middlewares by assigning to ``provider.middlewares``,
+before making the request, and then passes the response back up the middleware stack
+for processing. The next middleware on the stack is essentially the "make_request"
+method until we reach the end of the middlewares and the request is made by the actual
+`make_request` method of the provider. The response is then passed back up the
```
double-tick

pacrob commented on docs/internals.rst:
Line 13 says:
These can be *global* operating on all providers or specific to one provider.

I don't think this is still true - we don't have groups of providers anymore, right?

pacrob replied:
The diagram in *Request Lifecycle* shows both Global and Provider Middlewares

reedsa commented on docs/internals.rst:
```diff
@@ -117,66 +117,187 @@ You can set a new list of middlewares by assigning to ``provider.middlewares``,
+            return response
+
+
+.. code-block:: python
```
Extraneous code block
```suggestion
```

reedsa reviewed (approved):
Had one tiny suggestion but looks good.

fselmo committed:
changes from comments on PR #3211

fselmo commented:
@pacrob Thanks for the comments. Updated 👀 .

pacrob reviewed (approved):
lgtm!

kclowes commented on docs/examples.rst:
```diff
@@ -181,7 +181,7 @@ There are a few options for sending transactions:
 - :meth:`~web3.eth.Eth.send_raw_transaction`
 - Calling :meth:`~web3.contract.ContractFunction.transact` on a contract function
-- Utilizing :meth:`~web3.middleware.construct_sign_and_send_raw_middleware`
+- Configuring the ``build`` method for the :class:`~web3.middleware.SignAndSendRawMiddlewareBuilder`
```
nit: what do you think about something more human readable here? Maybe just `Configuring sign and send middleware` or something like that? With a link to the docs on how to do that

kclowes commented on docs/internals.rst:
```diff
@@ -117,66 +111,185 @@ You can set a new list of middlewares by assigning to ``provider.middlewares``,
 
-It is also possible to implement middlewares as a class.
 
+As with previous versions of the library, wrapping the ``make_request`` method of a
```
I would get rid of the 'As with previous versions of the library'. I don't think it's relevant here moving forward.

kclowes commented on docs/middleware.rst:
```diff
@@ -26,33 +20,20 @@ More information is available in the "Internals: :ref:`internals__middlewares`"
-* ``gas_estimate``
-
-Async middlewares include:
+The following middleware are added by default if you don't add any:
```
middlewares?

kclowes reviewed (approved):
Looks good to me! I left a few nits but nothing substantial. ⛵ I think we're slightly inconsistent with the use of middleware/middlewares for plural middleware, but I also don't think it matters too much. There were a few instances in the file I left the `middlewares?` comment on, but feel free to take it or leave it!

fselmo commented:
Sounds good, I've resolved all the outstanding questions here and also changed `_wrap_make_request` -> `wrap_make_request` for `Web3Middleware` classes since we instruct how to override that as well. It's an actually useful method to override and utilize so the underscore just adds confusion.

I also rebased to fix conflicts in tests.
